fud2/fud-core/src/exec/driver.rs

```diff
@@ -25,104 +24,176 @@ pub struct Driver {
+        start: &[StateRef],
+        end: &[StateRef],
         through: &[OpRef],
-    ) -> Option<Vec<OpRef>> {
-        let mut cur_state = start;
-        let mut op_path: Vec<OpRef> = vec![];
-
-        // Build path segments through each through required operation.
-        for op in through {
-            let segment =
-                self.find_path_segment(cur_state, Destination::Op(*op))?;
-            op_path.extend(segment);
-            cur_state = self.ops[*op].output;
-        }
-
-        // Build the final path segment to the destination state.
-        let segment =
-            self.find_path_segment(cur_state, Destination::State(end))?;
-        op_path.extend(segment);
+    ) -> Option<Vec<(OpRef, Vec<StateRef>)>> {
+        let good = |plan: &[(OpRef, Vec<StateRef>)]| {
+            let end_created = end.iter().all(|s| {
+                for (_, states) in plan {
+                    if states.contains(s) {
+                        return true;
+                    }
+                }
+                false
+            });
+            let through_used = through.iter().all(|t| {
+                for (op, used_states) in plan {
+                    if op == t {
+                        // TODO: figure out how to chain this so it check more than one level deep
+                        // required ops not being used
+                        return !used_states.is_empty();
+                    }
+                }
+                false
+            });
+            end_created && through_used
+        };
 
-        Some(op_path)
+        for len in 1..Self::MAX_PATH_LEN {
+            if let Some(plan) =
+                self.try_paths_of_length(&mut vec![], len, start, end, &good)
+            {
+                return Some(plan);
+            }
+        }
+        None
     }
 
     /// Generate a filename with an extension appropriate for the given State.
-    fn gen_name(&self, stem: &str, state: StateRef) -> Utf8PathBuf {
-        let state = &self.states[state];
-        if state.is_pseudo() {
-            Utf8PathBuf::from(format!("_pseudo_{}", state.name))
+    /// This function assumes all input states are unique and all output states are unique.
+    /// Additionally, it assume all non-input/output states are unique.
+    fn gen_name(
+        &self,
+        state_ref: StateRef,
+        used: bool,
+        req: &Request,
+        used_as_input: &mut HashSet<StateRef>,
+    ) -> IO {
+        let state = &self.states[state_ref];
+        let extension = if !state.extensions.is_empty() {
+            &state.extensions[0]
         } else {
-            // TODO avoid collisions in case we reuse extensions...
-            Utf8PathBuf::from(stem).with_extension(&state.extensions[0])
+            ""
+        };
+
+        // make sure we have correct input/output filenames and mark if we read from stdio
+        if req.start_state.contains(&state_ref)
+            && !used_as_input.contains(&state_ref)
```

Comment:
It was a bit of a hack which has since been removed. 

If you are still curious, the issue was I wanted a map from states to filenames which wasn't a function. There could be a state in both `req.start_state` and `req.end_state` which would then need two different files associated with it. `used_as_input` would get populated with this state the first time it was used as input to `gen_name` so `gen_name` would generate a different name on a subsequent call.